ports/esp32/machine_encoder.c

```diff
@@ -0,0 +1,787 @@
+STATIC mp_pcnt_obj_t *pcnts[PCNT_UNIT_MAX] = {};
+
+#if ESP_IDF_VERSION >= ESP_IDF_VERSION_VAL(4, 1, 0)
+#define EVT_THRES_0 PCNT_EVT_THRES_0
+#define EVT_THRES_1 PCNT_EVT_THRES_1
+#define EVT_ZERO    PCNT_EVT_ZERO
+#else
+#define EVT_THRES_0 (1 << PCNT_EVT_THRES_0)
+#define EVT_THRES_1 (1 << PCNT_EVT_THRES_1)
+#define EVT_ZERO    (1 << PCNT_EVT_ZERO)
+#endif
+
+/* Decode what PCNT's unit originated an interrupt
+ * and pass this information together with the event type
+ * the main program using a queue.
+ */
+#if CONFIG_IDF_TARGET_ESP32S2 || CONFIG_IDF_TARGET_ESP32S3
+#define H_LIM_LAT  cnt_thr_h_lim_lat_un
+#define L_LIM_LAT  cnt_thr_l_lim_lat_un
+#define THRES0_LAT cnt_thr_thres0_lat_un
+#define THRES1_LAT cnt_thr_thres0_lat_un
+#define ZERO_LAT   cnt_thr_zero_lat_un
+#else
+#define H_LIM_LAT  h_lim_lat
+#define L_LIM_LAT  l_lim_lat
+#define THRES0_LAT thres0_lat
+#define THRES1_LAT thres1_lat
+#define ZERO_LAT   zero_lat
+#endif
+STATIC void IRAM_ATTR pcnt_intr_handler(void *arg) {
+    for (int id = 0; id < PCNT_UNIT_MAX; ++id) {
+        if (PCNT.int_st.val & (1 << id)) {
+            mp_pcnt_obj_t *self = pcnts[id];
+            if (self != NULL) {
+                if (PCNT.status_unit[id].H_LIM_LAT) {
+                    self->counter += INT16_ROLL;
+                } else if (PCNT.status_unit[id].L_LIM_LAT) {
+                    self->counter -= INT16_ROLL;
+                }
+
+                self->status = 0;
+                if (PCNT.status_unit[id].THRES1_LAT) {
+                    if (self->counter == self->counter_match1) {
+                        self->status |= EVT_THRES_1;
+                        mp_sched_schedule(self->handler_match1, MP_OBJ_FROM_PTR(self));
+                        mp_hal_wake_main_task_from_isr();
+                    }
+                }
+                if (PCNT.status_unit[id].THRES0_LAT) {
+                    if (self->counter == self->counter_match2) {
+                        self->status |= EVT_THRES_0;
+                        mp_sched_schedule(self->handler_match2, MP_OBJ_FROM_PTR(self));
+                        mp_hal_wake_main_task_from_isr();
+                    }
+                }
+                if (PCNT.status_unit[id].ZERO_LAT) {
+                    if (self->counter == 0) {
+                        self->status |= EVT_ZERO;
+                        mp_sched_schedule(self->handler_zero, MP_OBJ_FROM_PTR(self));
+                        mp_hal_wake_main_task_from_isr();
+                    }
+                }
+            }
+            PCNT.int_clr.val |= 1 << id; // clear the interrupt
+        }
+    }
+}
+
+STATIC void register_isr_handler(void) {
+    if (pcnt_isr_handle == NULL) {
+        check_esp_err(pcnt_isr_register(pcnt_intr_handler, (void *)0, (int)0, (pcnt_isr_handle_t *)&pcnt_isr_handle));
+        if (pcnt_isr_handle == NULL) {
+            mp_raise_msg(&mp_type_Exception, MP_ERROR_TEXT("wrap interrupt failed"));
+        }
+    }
+}
```

Comment:
Using pcnt_isr_handler_add() instead would give the unit demultiplexing inside the isr handler for free, in turn simplifying the whole handler. See .e.g https://github.com/espressif/esp-idf/blob/faf0f61cdba67fa5f7e680f2f5b0966c8d69ce9a/examples/peripherals/pcnt/pulse_count_event/main/pcnt_event_example_main.c